File: backend/services/match_strategy.py
"""PRD 生成策略配置。"""

from dataclasses import dataclass, field
from typing import Any, Dict, Optional


@dataclass
class MatchStrategy:
    strict_mode: bool = True
    allow_cross_slot: bool = False
    dedup_policy: str = "global"
    prefer_4k: bool = True
    color_match_template: bool = True
    transition_inherit: bool = True
    use_vector_match: bool = True
    vector_weight: float = 0.25

    @classmethod
    def from_dict(cls, data: Optional[Dict[str, Any]]) -> "MatchStrategy":
        if not data:
            return cls()
        return cls(
            strict_mode=bool(data.get("strict_mode", True)),
            allow_cross_slot=bool(data.get("allow_cross_slot", False)),
            dedup_policy=str(data.get("dedup_policy", "global")),
            prefer_4k=bool(data.get("prefer_4k", True)),
            color_match_template=bool(data.get("color_match_template", True)),
            transition_inherit=bool(data.get("transition_inherit", True)),
            use_vector_match=bool(data.get("use_vector_match", True)),
            vector_weight=float(data.get("vector_weight", 0.25)),
        )

    def to_dict(self) -> Dict[str, Any]:
        return {
            "strict_mode": self.strict_mode,
            "allow_cross_slot": self.allow_cross_slot,
            "dedup_policy": self.dedup_policy,
            "prefer_4k": self.prefer_4k,
            "color_match_template": self.color_match_template,
            "transition_inherit": self.transition_inherit,
            "use_vector_match": self.use_vector_match,
            "vector_weight": self.vector_weight,
        }

    def merge_settings(self, settings: Optional[Dict[str, Any]]) -> Dict[str, Any]:
        base = {
            "strict_duration": self.strict_mode,
            "prefer_quality": self.prefer_4k,
            "dedup_policy": self.dedup_policy,
            "transition_inherit": self.transition_inherit,
            "use_vector_match": self.use_vector_match,
            "vector_weight": self.vector_weight,
        }
        if settings:
            base.update(settings)
        return base

File: frontend/lib/matchStrategy.ts
export type MatchStrategy = {
  strict_mode: boolean;
  allow_cross_slot: boolean;
  dedup_policy: 'global' | 'none';
  prefer_4k: boolean;
  color_match_template: boolean;
  transition_inherit: boolean;
  use_vector_match: boolean;
  vector_weight: number;
  semantic_weight: number;
  min_match_score: number;
};

export const DEFAULT_MATCH_STRATEGY: MatchStrategy = {
  strict_mode: true,
  allow_cross_slot: false,
  dedup_policy: 'global',
  prefer_4k: true,
  color_match_template: true,
  transition_inherit: true,
  use_vector_match: true,
  vector_weight: 0.25,
  semantic_weight: 0.4,
  min_match_score: 0.38,
};

export function strategyToSettings(strategy: MatchStrategy) {
  return {
    strict_duration: strategy.strict_mode,
    prefer_quality: strategy.prefer_4k,
    dedup_policy: strategy.dedup_policy,
    transition_inherit: strategy.transition_inherit,
    use_vector_match: strategy.use_vector_match,
    vector_weight: strategy.vector_weight,
    semantic_weight: strategy.semantic_weight,
    min_match_score: strategy.min_match_score,
  };
}
